src/Control/Effect/Cut.hs

```diff
@@ -0,0 +1,100 @@
+
+-- | Fail the current branch, and prevent backtracking within the nearest enclosing 'call' (if any).
+--
+--   Contrast with 'empty', which fails the current branch but allows backtracking.
+--
+--   prop> run (runNonDetOnce (runCut (cutfail <|> pure a))) == Nothing
+--   prop> run (runNonDetOnce (runCut (pure a <|> cutfail))) == Just a
+cutfail :: (Carrier sig m, Member Cut sig) => m a
+cutfail = send Cutfail
+
+-- | Delimit the effect of 'cutfail's, allowing backtracking to resume.
+--
+--   prop> run (runNonDetOnce (runCut (call (cutfail <|> pure a) <|> pure b))) == Just b
+call :: (Carrier sig m, Member Cut sig) => m a -> m a
+call m = send (Call m ret)
+
+-- | Commit to the current branch, preventing backtracking within the nearest enclosing 'call' (if any) on failure.
+--
+--   prop> run (runNonDetOnce (runCut (cut *> pure a <|> pure b))) == Just a
+--   prop> run (runNonDetOnce (runCut (cut *> empty <|> pure a))) == Nothing
+cut :: (Alternative m, Carrier sig m, Member Cut sig) => m ()
+cut = pure () <|> cutfail
+
+
+-- | The result of a nondeterministic branch of a computation.
+data Branch a
+  = Prune
+  | None
+  | Some a
+  deriving (Eq, Functor, Ord, Show)
+
+-- | Case analysis for 'Branch', taking a value to use for 'Prune', a value to use for 'None', and a function to apply to the contents of 'Some'.
+--
+--   prop> branch Prune None Some a == a
+--   prop> branch a b (applyFun f) Prune == a
+--   prop> branch a b (applyFun f) None == b
+--   prop> branch a b (applyFun f) (Some c) == applyFun f c
+branch :: a -> a -> (b -> a) -> Branch b -> a
+branch a _ _ Prune    = a
+branch _ a _ None     = a
+branch _ _ f (Some a) = f a
+
+
+-- | Run a 'Cut' effect within an underlying 'Alternative' instance (typically 'Eff' carrying a 'NonDet' effect).
+--
+--   prop> run (runNonDetOnce (runCut (pure a))) == Just a
+runCut :: (Alternative m, Carrier sig m, Effect sig, Monad m) => Eff (CutC m) a -> m a
+runCut = (>>= branch empty empty pure) . runCutC . interpret
+
+newtype CutC m a = CutC { runCutC :: m (Branch a) }
+
+instance (Alternative m, Carrier sig m, Effect sig, Monad m) => Carrier (Cut :+: NonDet :+: sig) (CutC m) where
+  ret = CutC . ret . Some
+  eff = CutC . handleSum (handleSum
+    (eff . handle (Some ()) (branch (ret Prune) (ret None) runCutC))
+    (\case
+      Empty    -> ret None
+      Choose k -> runCutC (k True) >>= branch (ret Prune) (runCutC (k False)) (\ a -> ret (Some a) <|> runCutC (k False))))
+    (\case
+      Cutfail  -> ret Prune
+      Call m k -> runCutC m >>= branch (ret None) (ret None) (runCutC . k))
+
+
+-- $setup
+-- >>> :seti -XFlexibleContexts
+-- >>> import Test.QuickCheck
+-- >>> import Control.Effect.Void
+-- >>> instance Arbitrary a => Arbitrary (Branch a) where arbitrary = frequency [(1, pure Prune), (1, pure None), (3, Some <$> arbitrary)] ; shrink b = case b of { Some a -> Prune : None : map Some (shrink a) ; None -> [Prune] ; Prune -> [] }
```

Comment:
I selected these frequencies based on the frequencies selected in the `Arbitrary1` instance for `Maybe`: `1` for `Nothing`, `3` for `Just`.